napalm_panos/panos.py

```diff
@@ -0,0 +1,326 @@
+# License. You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+
+# std libs
+import sys
+import xmltodict
+import json
+import pan.xapi
+import os.path
+import xml.etree
+import os
+import requests
+import requests_toolbelt
+from requests.packages.urllib3.exceptions import InsecureRequestWarning
+from datetime import datetime
+import time
+
+# local modules
+import napalm_base.exceptions
+import napalm_base.helpers
+from napalm_base.exceptions import ReplaceConfigException, MergeConfigException
+
+from napalm_base.base import NetworkDriver
+from netmiko import ConnectHandler
+
+
+class PANOSDriver(NetworkDriver):
+
+    def __init__(self, hostname, username, password, timeout=60):
+        self.hostname = hostname
+        self.username = username
+        self.password = password
+        self.timeout = timeout
+        self.loaded = False
+        self.changed = False
+        self.device = None
+        self.connected = False
+
+    def __enter__(self):
+        try:
+            self.open()
+        except:
+            exc_info = sys.exc_info()
+            self.__raise_clean_exception(exc_info[0], exc_info[1], exc_info[2])
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        self.close()
+        if exc_type is not None:
+            self.__raise_clean_exception(exc_type, exc_value, exc_traceback)
+
+    @staticmethod
+    def __raise_clean_exception(exc_type, exc_value, exc_traceback):
+        if exc_type.__name__ not in dir(napalm_base.exceptions) and \
+                        exc_type.__name__ not in __builtins__.keys():
+            epilog = ("NAPALM didn't catch this exception. Please, fill a bugfix on "
+                      "https://github.com/napalm-automation/napalm/issues\n"
+                      "Don't forget to include this traceback.")
+            print(epilog)
+        raise exc_type, exc_value, exc_traceback
+
+    def open(self):
+        self.device = pan.xapi.PanXapi(hostname=self.hostname,
+                                       api_username=self.username,
+                                       api_password=self.password)
+
+    def _open_ssh(self):
+        self.connected = ConnectHandler(device_type='paloalto_panos',
+                                        ip=self.hostname,
+                                        username=self.username,
+                                        password=self.password)
+
+    def close(self):
+        self.device = None
+        self.connected.disconnect()
+
+    def load_template(self, template_name, **template_vars):
```

Comment:
And anyway we are doing some changes right now: https://github.com/napalm-automation/napalm-base/pull/27